### PR TITLE
x/gov: add ElectionRule address and pregernate ID

### LINKED ISSUE #644: Persist escrow/ aswap/ distribution address in model
**Is your feature request related to a problem? Please describe.**

*Very useful to allow cosmos station to connect without using iov-core.*

The following modules generate addresses to store coins and transfer them based on their smart contract rules:
- [escrow ](https://github.com/iov-one/weave/blob/master/x/escrow/model.go#L102)
- [aswap ](https://github.com/iov-one/weave/blob/master/x/aswap/handler.go#L286-L288)
- [distribution](https://github.com/iov-one/weave/blob/master/x/distribution/model.go#L119-L123)
- gov (ElectionRule)
- multisig (Contract)
- paychan

The address calculation is in code only and not stored with the model data. This moves the complexity for recalculating the address to the client side to query a balance. Backend logic is replicated in the client (iov-core).
By storing the address within the model we also gain extra safety that no coins ever get locked due to changes in the address calculation rules.

**Describe the solution you'd like**
Store the generated address within the model so that it will be returned on queries. So that it can be used for querying the balance without further processing.

**Additional context**
iov-core can calculate addresses based on conditions.

Note that we need to have an ID before calculating an address. On creation, we only get the ID when Saving, so we would need to save twice, or add some callback to set Address in the model in a cleaner way.

We need to design the how - but at least they what and why are clear

### PR DESCRIPTION
**Pull request by ruseinov**
Potentially fixes #644

### CODE REVIEW

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: x/gov/codec.proto [hunk 1]
```diff
@@ -60,6 +60,8 @@ message ElectionRule {
   // The valid range for the threshold value is `0.5` to `1` (inclusive) which allows any value between half and all
   // of the eligible voters.
   Fraction quorum = 8;
+  // Address of this entity. Set during creation and does not change.
+  bytes address = 9 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
```
Comment (alpe) [inline x/gov/codec.proto]: 👍

File: x/gov/init.go [hunk 2]
```diff
@@ -72,27 +72,35 @@ func (*Initializer) FromGenesis(opts weave.Options, params weave.GenesisParams,
...
+		}
+
 		rule := ElectionRule{
 			Metadata:     &weave.Metadata{Schema: 1},
 			Admin:        r.Admin,
 			Title:        r.Title,
 			VotingPeriod: r.VotingPeriod,
 			Threshold:    Fraction{Numerator: r.Threshold.Numerator, Denominator: r.Threshold.Denominator},
 			ElectorateID: electorateID,
+			Address: weave.Condition(newRuleID).Address(),
```
Comment (alpe) [inline x/gov/init.go]: req: you can not cast to condition but create a new one via `NewCondition` with [`gov/rule/newRuleID`](https://github.com/iov-one/devnet-operations/blob/master/manifests/weave/davenet/bns/genesis.json#L136)

File: x/gov/model.go [hunk 3]
```diff
@@ -135,6 +135,9 @@ func (m ElectionRule) Validate() error {
 			return errors.Wrap(err, "quorum")
 		}
 	}
+	if err := m.Address.Validate(); err != nil {
```
Comment (alpe) [inline x/gov/model.go]: 👍

File: orm/versioning_bucket.go [hunk 4]
```diff
@@ -107,18 +107,28 @@ func (b VersioningBucket) GetVersion(db weave.ReadOnlyKVStore, ref VersionedIDRe
...
 		return nil, err
 	}
-	idRef := VersionedIDRef{ID: newID, Version: data.GetVersion()}
+	return b.create(db, newID, data)
+}
+
+// CreateWithID stores the given data. It accepts an ID and assigns an initial version number to the object instance
+// and returns the VersionedIDRef which won't be nil on success.
+func (b VersioningBucket) CreateWithID(db weave.KVStore, id []byte, data versionedData) (*VersionedIDRef, error) {
+	return b.create(db, id, data)
```
Comment (alpe) [inline orm/versioning_bucket.go]: req: at the framework level we should prevent an id to be nil or empty

File: orm/versioning_bucket.go [hunk 5]
```diff
@@ -107,18 +107,28 @@ func (b VersioningBucket) GetVersion(db weave.ReadOnlyKVStore, ref VersionedIDRe
...
 	if err != nil {
 		return nil, err
 	}
-	idRef := VersionedIDRef{ID: newID, Version: data.GetVersion()}
+	return b.create(db, newID, data)
+}
+
+// CreateWithID stores the given data. It accepts an ID and assigns an initial version number to the object instance
+// and returns the VersionedIDRef which won't be nil on success.
+func (b VersioningBucket) CreateWithID(db weave.KVStore, id []byte, data versionedData) (*VersionedIDRef, error) {
```
Comment (alpe) [inline orm/versioning_bucket.go]: pp: I can see why you want the ID as an argument but the `WithVersioning(b IDGenBucket) VersioningBucket` constructor requires an IDGenBucket which is fully ignored by this method now. 

Why not drop that dependency and the other `Create` method and always require an ID argument? The only other place beside tests that uses the method is the electorate Bucket.

File: x/gov/bucket.go [hunk 6]
```diff
@@ -50,7 +52,7 @@ type ElectionRulesBucket struct {
 func NewElectionRulesBucket() *ElectionRulesBucket {
 	b := migration.NewBucket(packageName, "electnrule", orm.NewSimpleObj(nil, &ElectionRule{}))
 	return &ElectionRulesBucket{
-		VersioningBucket: orm.WithVersioning(orm.WithSeqIDGenerator(b, "id")),
+		VersioningBucket: orm.WithVersioning(orm.WithSeqIDGenerator(b, electionRuleSequence)),
```
Comment (alpe) [inline x/gov/bucket.go]: pp: as I wrote in my other comment, better drop the sequence here than being inconsistent in the api

Comment (ruseinov) [reply x/gov/bucket.go]: Here's the motivation behind this - if we ever want to use this bucket for anything else - we'll be forced to use manual id generation.

Since current implementations both use the same sequence - I'd say it is not such a bad thing to support both, but I do understand the idea.

WDYT?

There might be a better solution to this problem too, but I don't want it to drag along.

File: orm/versioning_bucket.go (hunk 5 above)
Comment (ruseinov) [reply orm/versioning_bucket.go]: Main reason is that `CreateWithID` is basically used to cater for the Address edge-case, in other instances it's nice to have an autogenerated id.

Since they are referencing the same sequence - I'd say it should be fine to have both.
There might be a slick(er) solution for this actually, but that'd require using interfaces instead of structs and a bigger change.

**Comment by ruseinov** [comment]
@alpe As discussed in PR comments - let's keep both `CreateWithID` and `Create` methods as `CreateWithID` method is supposed to cater for an edge-case.

I will, however, add a comment to it to make sure it's not used when not needed.

WDYT?

**Comment by ruseinov** [comment]
Proceeding with addressing review comments and tests for now.

**Comment by ruseinov** [comment]
@alpe Addressed requirements and introduced tests + a little helper method in `ElectionRulesBucket` to aid testing (basically NextID).

File: x/gov/bucket.go [hunk 7]
```diff
@@ -50,10 +52,15 @@ type ElectionRulesBucket struct {
...
 	}
 }
 
+func (b *ElectionRulesBucket) NextID(db weave.KVStore) ([]byte, error) {
```
Comment (alpe) [inline x/gov/bucket.go]: 👍
